File: Cpp/Longest Common Subsequence.cpp
class Solution {
public:
    /**
     * @param A, B: Two strings.
     * @return: The length of longest common subsequence of A and B.
     */
    int longestCommonSubsequence(string A, string B) {
        // write your code here
        int m = A.size(), n = B.size();
        vector<vector<int> > vec(m+1, vector<int>(n+1));
        vec[0][0] = 0;
        for (int i = 0; i < m; i ++) {
            for (int j = 0; j < n; j ++) {
                if (A[i] == B[j]) {
                    vec[i+1][j+1] = vec[i][j] + 1;
                } else {
                    vec[i+1][j+1] = max(vec[i][j+1], vec[i+1][j]);
                }
            }
        }
        return vec[m][n];
    }
};

File: Java/Longest-Common-Subsequence.java
public class Solution {
    /**
     * @param A, B: Two strings.
     * @return: The length of longest common subsequence of A and B.
     */
    public int longestCommonSubsequence(String A, String B) {
        // write your code here
        int lenA = A.length();
        int lenB = B.length();
        if (lenA == 0 || lenB == 0) return 0;
        int [][]lcs = new int[lenA + 1][lenB + 1];
        lcs[0][0] = 0;
        for (int i = 1; i <= lenA; i ++) {
        	for (int j = 1; j <= lenB; j ++) {
        		if (A.charAt(i - 1) == B.charAt(j - 1)) {
        			lcs[i][j] = lcs[i - 1][j - 1] + 1;
        		} else {
        			lcs[i][j] = Math.max(lcs[i - 1][j], lcs[i][j - 1]);
        		}
        	}
        }
        return lcs[lenA][lenB];
    }
}

